ocaml/xapi/xapi_vm_migrate.ml

```diff
@@ -980,26 +1007,34 @@ let vdi_copy_fun __context dbg vdi_map remote is_intra_pool remote_vdis so_far
   let mirror_to_remote new_dp =
     let task =
       if not vconf.do_mirror then
-        SMAPI.DATA.copy dbg vconf.sr vconf.location remote.sm_url dest_sr
-          is_intra_pool
+        SMAPI.DATA.copy dbg vconf.sr vconf.location vconf.copy_vm remote.sm_url
+          dest_sr is_intra_pool
       else
         (* Though we have no intention of "write", here we use the same mode as the
            associated VBD on a mirrored VDIs (i.e. always RW). This avoids problem
            when we need to start/stop the VM along the migration. *)
         let read_write = true in
         (* DP set up is only essential for MIRROR.start/stop due to their open ended pattern.
            It's not necessary for copy which will take care of that itself. *)
-        let vm = Storage_interface.Vm.of_string "0" in
         ignore
-          (SMAPI.VDI.attach3 dbg new_dp vconf.sr vconf.location vm read_write) ;
-        SMAPI.VDI.activate dbg new_dp vconf.sr vconf.location ;
+          (SMAPI.VDI.attach3 dbg new_dp vconf.sr vconf.location vconf.mirror_vm
+             read_write
+          ) ;
+        SMAPI.VDI.activate3 dbg new_dp vconf.sr vconf.location vconf.mirror_vm ;
         let id =
           Storage_migrate.State.mirror_id_of (vconf.sr, vconf.location)
         in
+        debug "%s mirror_vm is %s copy_vm is %s" __FUNCTION__
+          (Vm.string_of vconf.mirror_vm)
+          (Vm.string_of vconf.copy_vm) ;
         (* Layering violation!! *)
         ignore (Storage_access.register_mirror __context id) ;
-        SMAPI.DATA.MIRROR.start dbg vconf.sr vconf.location new_dp remote.sm_url
-          dest_sr is_intra_pool
+        (* XXX I really do not understand why we need to pass copy_vm and then
+           mirror_vm. The storage_interface defines mirror_vm first. But logging
+           suggests that Storage_mux receives copy_vm first so I had to reverse the ordering
+           of these two params to make it work. *)
+        SMAPI.DATA.MIRROR.start dbg vconf.sr vconf.location new_dp vconf.copy_vm
```

Comment:
If anyone understands what is going on here, please let me know :)

Reply:
Doesn't make any sense at all... Can you use the labelled arguments, for everyone's sanity?